test/models/streamed/state.spec.ts

```diff
@@ -0,0 +1,203 @@
+import { expect } from "chai"
+import { BatchExchange } from "../../../src"
+import { DEFAULT_ORDERBOOK_OPTIONS } from "../../../src/streamed"
+import { AuctionState } from "../../../src/streamed/state"
+import { AnyEvent, EventName, EventValues } from "../../../src/streamed/events"
+
+function auctionState(): AuctionState {
+  return new AuctionState(
+    { ...DEFAULT_ORDERBOOK_OPTIONS, strict: true },
+  )
+}
+
+type Named<T> = Pick<T, keyof T & string>
+
+function event<
+  K extends EventName<BatchExchange>,
+  V extends BatchExchange["events"][K],
+>(
+  block: number,
+  name: K,
+  data:  Named<EventValues<V>>,
+  index: number = 0,
+): AnyEvent<BatchExchange> {
+  // NOTE: Cast to `any` as there are missing event data properties that the
+  // account state doesn't care about.
+  return {
+    event: name,
+    blockNumber: block,
+    returnValues: data,
+    logIndex: index || 0,
+  } as any
+}
+
+function addr(lowerBits: number): string {
+  return `0x${lowerBits.toString(16).padStart(40, "0")}`
+}
+
+describe("Account State", () => {
+  describe("applyEventsUntilBatch", () => {
+    it("Updates the batch ID", async () => {
+      const state = auctionState()
+      state.applyEvents([
+        event(0, "TokenListing", { id: "0", token: addr(1) }),
+      ])
+      expect(state.nextBlock).to.equal(1)
+    })
+
+    it("Throws when past block is applied", async () => {
+      const state = auctionState()
+      state.applyEvents([
+        event(10, "TokenListing", { id: "0", token: addr(1) }),
+      ])
+      expect(() => state.applyEvents([
+        event(1, "TokenListing", { id: "1", token: addr(1) }),
+      ])).to.throw()
+    })
+
+    it("Throws when events are out of order", async () => {
+      const state = auctionState()
+      expect(() => state.applyEvents([
+        event(2, "TokenListing", { id: "0", token: addr(1) }),
+        event(1, "TokenListing", { id: "0", token: addr(1) }),
+      ])).to.throw()
+    })
+  })
+
+  describe("OrderPlacement > OrderCancellation > OrderDeletion", () => {
```

Comment:
> @fleupold Nit: no strong feelings, but this could also be three test cases as unit tests should usually only test one thing.

I completely agree, I just didn't know the best way to do it since the `OrderDeletion` test's setup will be pretty much identical to the `OrderCancellation` test, and the `OrderCancellation` test's setup will be pretty much the same as the `OrderPlacement` test.

Basically, `OrderDeletion` tests the `OrderPlacement` and `OrderCancellation` as well. Do you have any suggestions or should we just duplicate the setup and :see_no_evil:?

Reply:
I don't necessarily think this is better here, but I'd probably define the order event as a const outside and then have three small test cases (duplicating the TokenListing/OrderPlacement code). But up to you...

Reply:
I really like that idea!

Reply:
I was actually being silly, I just separated into 3 separate tests and they are still very readable, so my initial concerns weren't really warranted.